### PR TITLE
feat: Add support for in-app classification

### LINKED ISSUE #472: Support InAppIncludes
**Description**

Support InAppIncludes for Android and mark frames as inApp if part of the frame's function or package

**When does the problem happen**

Frames are always inApp=false on Android events and grouping is suboptimal

- [ ] During build
- [X] During run-time
- [ ] When capturing a hard crash

**Environment**

any Android version

**Steps To Reproduce**

Just capture an event and see `in_app` in the raw JSON

### LINKED ISSUE #473: Extend Sentry Options to better integrate with Android
When initializing `sentry-react-native` or `sentry-flutter` the user provides some options such as:

* maxBreadcrumb
* inAppInclude
* ...

These SDKs in turn call `sentry-android`'s init, propagating the values from the options. Not everything is supported, but many options are.

Between `sentry-android` and `sentry-native` not many options are supported.

We'd like to extend the support between these two SDKs.

Relates to #472

The options we have in [Java](https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/SentryOptions.java) and [Android](https://github.com/getsentry/sentry-java/blob/main/sentry-android-core/src/main/java/io/sentry/android/core/SentryAndroidOptions.java)

### PR DESCRIPTION
**Pull request by Swatinem**
This needed some changes to the way that options are passed through when applying the scope to events.

Fixes #472
Fixes #473 (we already have `max_breadcrumbs`, and `close` is being added in #518)

### CODE REVIEW

File: include/sentry.h [hunk 1]
```diff
@@ -834,12 +834,23 @@ SENTRY_API void sentry_options_set_symbolize_stacktraces(
...
 
+/**
+ * Adds a new pattern for *in-app* classification.
+ *
+ * When on-device symbolication is enabled (see
+ * `sentry_options_set_symbolize_stacktraces`), this will flag stack frames as
+ * *in-app*, which match this pattern.
+ * Matching is done on both symbols names and object file names they belong to.
+ */
+SENTRY_API void sentry_options_add_in_app_include(
```
Comment (jan-auer) [inline include/sentry.h:845]: Please also add `in_app_exclude`. After merging, let's add `native` to supported platforms here:

https://github.com/getsentry/sentry-docs/blob/322a46a6bd98f7f564db74e8d7752312d94c3cee/src/platforms/common/configuration/options.mdx#L157-L171

File: include/sentry.h [hunk 2]
```diff
@@ -834,12 +834,23 @@ SENTRY_API void sentry_options_set_symbolize_stacktraces(
 SENTRY_API int sentry_options_get_symbolize_stacktraces(
     const sentry_options_t *opts);
 
+/**
+ * Adds a new pattern for *in-app* classification.
+ *
+ * When on-device symbolication is enabled (see
+ * `sentry_options_set_symbolize_stacktraces`), this will flag stack frames as
+ * *in-app*, which match this pattern.
+ * Matching is done on both symbols names and object file names they belong to.
```
Comment (jan-auer) [inline include/sentry.h:843]: "Pattern" here is underspecified. If you compare with existing SDKs, such as JavaScript [documented here](https://docs.sentry.io/platforms/python/configuration/options/#in-app-include), this is only a prefix match on the module.

@untitaker @mitsuhiko - did we never support functions for this? Being able to do a prefix match on functions sounds pretty essential, tbh.

File: src/sentry_scope.c [hunk 3]
```diff
@@ -206,10 +214,28 @@ sentry__symbolize_frame(const sentry_frame_info_t *info, void *data)
...
     }
+
+    const char *symbol_name
+        = sentry_value_as_string(sentry_value_get_by_key(frame, "function"));
+    const char *object_name
+        = sentry_value_as_string(sentry_value_get_by_key(frame, "package"));
+
+    sentry_string_list_t *in_app = symbolize_data->options->in_app_includes;
+    while (in_app) {
+        if (strstr(symbol_name, in_app->str) != NULL
```
Comment (jan-auer) [inline src/sentry_scope.c:225]: With a prefix match, you can use `strncmp` here.

File: include/sentry.h (hunk 2 above)
Comment (untitaker) [reply include/sentry.h:843]: prefix match on the module was entirely sufficient for python. function names of a library don't share a common pattern one could match against

istm native just needs a different API, but that's fine IMO

**Pull request closed without merging**